R/expect_null_linter.R

```diff
@@ -0,0 +1,48 @@
+#' expect_null Linter
+#'
+#' Require usage of expect_null(x) over expect_equal(x, NULL) and similar
+#'   [testthat::expect_null()] exists specifically for testing for `NULL` objects.
+#'   [testthat::expect_equal()], [testthat::expect_identical()], and
+#'   [testthat::expect_true()] can also be used for such tests,
+#'   but it is better to use the tailored function instead.
+#'
+#' @evalRd rd_tags("expect_null_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+expect_null_linter <- function() {
+  Linter(function(source_file) {
+    if (length(source_file$parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    # for expect_{equal,identical}(x, NULL)
+    equal_expr_cond <- sprintf(
+      "SYMBOL_FUNCTION_CALL[%s] and following-sibling::expr[2][NULL_CONST]",
+      xp_text_in_table(c("expect_equal", "expect_identical"))
+    )
+
+    # for expect_true(is.null(x))
+    is_null_call <- "SYMBOL_FUNCTION_CALL[text() = 'is.null']"
+    true_expr_cond <- xp_and(
+      "SYMBOL_FUNCTION_CALL[text() = 'expect_true']",
+      sprintf("following-sibling::expr[1][expr[%s]]", is_null_call)
```

Comment:
I believe we don't need the `[1]` qualifier.

Also, `xp_and()` could be used above as well, no?

Reply:
re: `xp_and()`, recently I'm refactor to use a literal xpath as much as possible, I think it maximizes readability. done in the current commit.

I think the addition of `glue()` really enhances readability but we could try and stick with `sprintf()` to avoid the dep, WDYT?

Reply:
(it may also be good to just write out `text() = 'expect_equal' or text() = 'expect_identical'` here too, but there will be other linters to import that face the same question in better use cases)

Reply:
the `[1]` qualifier is for `expect_true(x, info = is.null(y))`. Maybe unlikely but doesn't hurt much to include it.

(the more general issue of really agreeing with `match.call()` is ignored. i think that it's typically OK to throw a false positive if people are using really unexpected positional arguments)